zerver/tests/test_push_registration.py

```diff
@@ -190,3 +197,192 @@ def test_register_push_device_error(self) -> None:
+        self.login("hamlet")
+
+        push_devices_count = PushDevice.objects.count()
+        self.assertEqual(push_devices_count, 0)
+
+        payload = self.get_register_push_device_payload()
+
+        # Verify the created PushDevice row while the registration
+        # to the bouncer is in progress (worker hasn't consumed yet).
+        with mock.patch("zerver.views.push_notifications.queue_event_on_commit"):
+            result = self.client_post("/json/mobile_push/register", payload)
+        self.assert_json_success(result)
+        push_devices = PushDevice.objects.all()
+        self.assert_length(push_devices, 1)
+        self.assertIsNone(push_devices[0].bouncer_device_id)
+        self.assertEqual(push_devices[0].status, "pending")
+
+        # Verify the created PushDevice row after the
+        # registration to the bouncer is complete.
+        with self.capture_send_event_calls(expected_num_events=1) as events:
+            payload["push_account_id"] = 9999
+            result = self.client_post("/json/mobile_push/register", payload)
+        self.assert_json_success(result)
+        push_devices = PushDevice.objects.order_by("pk")
+        self.assert_length(push_devices, 2)
+        self.assertIsNotNone(push_devices[1].bouncer_device_id)
+        self.assertEqual(push_devices[1].status, "active")
+        self.assertEqual(
+            events[0]["event"],
+            dict(type="push_device", op="update", push_account_id="9999", status="active"),
+        )
+
+        # Idempotent
+        with self.capture_send_event_calls(expected_num_events=0):
+            result = self.client_post("/json/mobile_push/register", payload)
+        self.assert_json_success(result)
+        push_devices = PushDevice.objects.all()
+        self.assert_length(push_devices, 2)
+
+        # For coverage
+        with (
+            self.settings(ZILENCER_ENABLED=False),
+            mock.patch(
+                "zerver.lib.push_registration.send_to_push_bouncer",
+                return_value={"device_id": 4},
+            ),
+            self.capture_send_event_calls(expected_num_events=1),
+        ):
+            payload["push_account_id"] = 5555
+            result = self.client_post("/json/mobile_push/register", payload)
+        self.assert_json_success(result)
+        push_devices = PushDevice.objects.all()
+        self.assert_length(push_devices, 3)
+
+    def test_register_push_device_error(self) -> None:
+        hamlet = self.example_user("hamlet")
+        self.login_user(hamlet)
+
+        push_devices_count = PushDevice.objects.count()
+        self.assertEqual(push_devices_count, 0)
+
+        payload = self.get_register_push_device_payload()
+
+        # Verify InvalidBouncerPublicKeyError
+        with (
+            mock.patch(
+                "zerver.lib.push_registration.do_register_remote_push_device",
+                side_effect=InvalidBouncerPublicKeyError,
+            ),
```

Comment:
I don't like doing this mocking strategy. We should be just sending an invalid bouncer public key, so the test verifies that it flows through the system end-to-end properly. 

https://zulip.readthedocs.io/en/stable/testing/philosophy.html#integration-testing-or-unit-testing has some discussion.